### PR TITLE
[WIP] Message Archive Management (XEP-0313)

### LINKED ISSUE #150: See complete history (XEP-0313)
I really like your app, but there is one thing missing.
I already compiled and installed a new version of ejabberd, to get mod_carboncopy ([XEP-0280](http://xmpp.org/extensions/xep-0280.html)) and mod_mam ([XEP-0313](http://xmpp.org/extensions/xep-0313.html))

I also activated Carbon Copy in jsxc-owncloud Settings.
But I have two problems:
- I can only access the last ~10 messages and not the whole history
- The history of my laptop is only on my laptop and not on my desktop pc

And idea? :)
##

### PR DESCRIPTION
**Pull request by singh1114**
fix #150 
Ref: http://xmpp.org/extensions/xep-0313.html

### CODE REVIEW

File: src/jsxc.lib.gui.js [hunk 1]
```diff
@@ -2360,6 +2360,7 @@ jsxc.gui.window = {
     */
    open: function(bid) {
       var win = jsxc.gui.window.init(bid);
+      jsxc.xmpp.mam.getHistory(bid);
```
Comment (sualko) [inline src/jsxc.lib.gui.js]: Please use an event handler for this. I think the appropriate event would be `init.window.jsxc`.

File: src/jsxc.lib.xmpp.mam.js [hunk 2]
```diff
@@ -27,14 +27,111 @@
...
+
+  //self.getHistory(userid, count, before);
+};
+
+
+jsxc.xmpp.mam.getHistory = function(bid) {
+  /*
+  To include the whole functionality of strophe-plugin we used self
+  */
+  var self = jsxc.xmpp.conn.mam;
```
Comment (sualko) [inline src/jsxc.lib.xmpp.mam.js]: `self` usually points to the home namespace. In this case it would be `jsxc.xmpp.mam`.

Comment (sualko) [inline src/jsxc.lib.xmpp.mam.js]: You have to make sure, that the connection has already be established. See e.g. [jsxc.lib.xmpp.chatState.js#L17](https://github.com/jsxc/jsxc/blob/master/src/jsxc.lib.xmpp.chatState.js#L17)

File: src/jsxc.lib.xmpp.mam.js [hunk 3]
```diff
@@ -27,14 +27,111 @@
...
+        <no-store xmlns="urn:xmpp:hints"></no-store>
+      </message>
+      
+      */
+      var direction;
+
+      /*
+      Check the direction of the message i.e. In which direction messages are exchanged.
+      */
+      if(bid === $(message).find("forwarded message").attr("to")){
```
Comment (sualko) [inline src/jsxc.lib.xmpp.mam.js]: Please use also the xml namespace to identify the forwarded element. See [jsxc.lib.xmpp.js#L789](https://github.com/jsxc/jsxc/blob/master/src/jsxc.lib.xmpp.js#L789).

File: src/jsxc.lib.xmpp.mam.js [hunk 4]
```diff
@@ -27,14 +27,111 @@
...
+        "type": "plain",
+        "bid": bid,
+        "direction": direction,
+        "msg": msgBody
+      };
+
+      //console.log(message);
+      //console.log(mamUid);
+
+      jsxc.storage.setUserItem('msg', mamUid, msg);
```
Comment (sualko) [inline src/jsxc.lib.xmpp.mam.js]: This shouldn't be needed.

File: src/jsxc.lib.xmpp.mam.js (hunk 2 above)
Comment (singh1114) [inline src/jsxc.lib.xmpp.mam.js]: Should I use 'this' than?

File: src/jsxc.lib.xmpp.mam.js (hunk 4 above)
Comment (singh1114) [inline src/jsxc.lib.xmpp.mam.js]: OK. This will be removed.

File: src/jsxc.lib.xmpp.mam.js (hunk 3 above)
Comment (singh1114) [inline src/jsxc.lib.xmpp.mam.js]: This is always equal to "<forwarded xmlns="urn:xmpp:forward:0">". What do we need to check here?

File: src/jsxc.lib.xmpp.mam.js (hunk 2 above)
Comment (sualko) [inline src/jsxc.lib.xmpp.mam.js]: ???

File: src/jsxc.lib.xmpp.mam.js (hunk 3 above)
Comment (sualko) [inline src/jsxc.lib.xmpp.mam.js]: In the future there could be an element with the same name, but different behavior.

**Comment by singh1114** [comment]
I can't get anything on the sorting of the messages. Please help!

**Comment by sualko** [comment]
If there is no sorting you have to implement it.

**Comment by singh1114** [comment]
I have a doubt up here. I know the postMessage function can handle a single message at a time. So, how can I sort a single message?( Please tell me if I am wrong) 

Also, used a logic like this....

I tweaked the code of getHistory function such that it can whenever a user clicked on the someone from the roster list, all the messages are stored in a list and at the end, they are transferred to the postMessage function.

I wasn't able to get the required output but I want to ask is it the right way of solving the problem?

Another way is to create a function for sorting and when all the messages are stored in a data structure by the function getHistory they can be passed to that sorting function and finally they can be passed one by one to the postMessage. 

Also if someone can use the code written by me and see if the messages of MUC were retrieved or not it would be great.

**Comment by sualko** [comment]
I would tweak the `_postMessage` function to work in a similar fashion than `jsxc.gui.roster.insert`.

**Comment by skyfox675** [comment]
Quick feature question: Are there plans to add an option either in settings (global) or on a per conversation/chat basis to set history (On | Off | Server Default | Recipient Setting)? This is a very nice feature, not sure the difficult required to implement, just thought I would ask.

Thanks!

**Comment by sualko** [comment]
@skyfox675 Thanks for your request. We will keep this in mind.

**Comment by singh1114** [comment]
@sualko Can you check the sortMessages function. Have I done it correctly?

File: src/jsxc.lib.gui.js [hunk 5]
```diff
@@ -2644,20 +2644,57 @@ jsxc.gui.window = {
...
+	            direction: jsxc.Message.SYS,
+	            msg: '42'
+	         }));
          }
       }
+      return message;
+   },
+
+   sortMessages: function(message){
+      var messageArray = jsxc.options.allMessages;
```
Comment (sualko) [inline src/jsxc.lib.gui.js]: Why do you save messages in the options object?

File: src/jsxc.lib.gui.js [hunk 6]
```diff
@@ -2644,20 +2644,57 @@ jsxc.gui.window = {
...
+      var length = messageArray.length;
+      var position = length;
+      var firstMessage;
 
+      do{
+         if (typeof messageArray[position-1] === 'undefined') {
+            firstMessage = true;
+         }
+         if (isNaN(message.stamp)) {
+            length = messageArray.length;
```
Comment (sualko) [inline src/jsxc.lib.gui.js]: ???

File: src/jsxc.lib.options.js [hunk 7]
```diff
@@ -5,6 +5,12 @@
  */
 jsxc.options = {
 
+   /**
+    * Declare empty list to be used for sorting the messages 
+    *  
+   */
+   allMessages: [],
```
Comment (sualko) [inline src/jsxc.lib.options.js]: The options object is only for options and not arbitrary data.

**Comment by sualko** [comment]
> Have I done it correctly?

I don't think so. It would be probably more elegant if you would handle the sorting at [jsxc.lib.gui.js#L2803](https://github.com/jsxc/jsxc/blob/master/src/jsxc.lib.gui.js#L2803).

We also have to think about delayed messages, because with the new sorting function these will be maybe not visible to the user.

File: src/jsxc.lib.gui.js (hunk 5 above)
Comment (singh1114) [inline src/jsxc.lib.gui.js]: I was trying to do it in some other ways. For that, I initialized an empty data structure in a function. So whenever the function ran it again got initialized to empty.

For example: say I am declaring a variable in sortMessage function.
`sortMessages: function(message) {
    var messageArray = []
}`
So whenever the function ran it again got initialized to empty deleting all the messages. 

If you have any other idea please share.

File: src/jsxc.lib.gui.js (hunk 6 above)
Comment (singh1114) [inline src/jsxc.lib.gui.js]: So when I working with the messages I found that the last incoming message is not a real message it is just a system message which says that all messages have been encountered. I also found that the stamp of the last message returns NaN. So I am using that to check few things.

File: src/jsxc.lib.options.js (hunk 7 above)
Comment (singh1114) [inline src/jsxc.lib.options.js]: Please suggest something.

File: src/jsxc.lib.gui.js (hunk 5 above)
Comment (sualko) [inline src/jsxc.lib.gui.js]: There has to be another solution, because this is :-1:.

File: src/jsxc.lib.gui.js (hunk 6 above)
Comment (sualko) [inline src/jsxc.lib.gui.js]: I mean, length was defined a couple of lines above and I think it is never changed.

**Comment by sualko** [comment]
Maybe you should start with some simpler tasks, like our new [starter issues](https://github.com/jsxc/jsxc/labels/starter), to get familiar with the code.

**Comment by singh1114** [comment]
How about #341...

**Comment by sualko** [comment]
I think we close this pr. Thanks for your effort.

**Pull request closed without merging**